chat-plugins/modlog.js

```diff
@@ -145,12 +137,65 @@ function getMoreButton(room, search, useExactSearch, lines, maxLines) {
 	if (!newLines || lines < maxLines) {
 		return ''; // don't show a button if no more pre-set increments are valid or if the amount of results is already below the max
 	} else {
-		if (useExactSearch) search = Chat.escapeHTML(`"${search}"`);
 		return `<br /><div style="float:right"><button class="button" name="send" value="/modlog ${room}, ${search} ${LINES_SEPARATOR}${newLines}" title="View more results">More results...</button></div>`;
 	}
 }
 
-async function runModlog(rooms, searchString, exactSearch, maxLines) {
+function toModlogId(text) {
+	if (typeof text !== 'string' && typeof text !== 'number') return '';
+	return ('' + text).toLowerCase().replace(/[^a-z0-9*]+/g, '').replace(/\*/g, '.*');
+}
+
+function formatSingleSearch(string) {
+	const searchType = string.trim().charAt(0) === '!' ? 'negative' : 'positive';
+	const parts = string.split('|');
+	let formattedParts = [];
+
+	for (let p of parts) {
+		p = p.trim();
+		const exactSearch = /^(["'`]).+\1$/i.test(p);
+		if (exactSearch) {
+			p = p.slice(1, -1);
+			formattedParts.push(p.replace(/[\\.+*?()|[\]{}^$]/g, '\\$&'));
+		} else {
+			formattedParts.push(`[^a-zA-Z0-9]${toModlogId(p).replace(/[a-z0-9]/g, '$&[^a-zA-Z0-9]*')}[^a-zA-Z0-9]`);
+		}
+	}
+
+	const result = {searchType, term: (formattedParts.length > 1 ? `(${formattedParts.join('|')})` : formattedParts[0])};
+	return result;
+}
+
+function formatRegexString(string) {
+	const parts = string.split(',');
+	if (parts.length === 1 && string.charAt(0) !== '!' && !string.includes('|')) {
+		// single argument searches
+		if (/^(["'`]).+\1$/i.test(string)) {
+			// exact search
+			return {type: 1, searchRegex: string.slice(1, -1).replace(/[\\.+*?()|[\]{}^$]/g, '\\$&')};
```

Comment:
This replacement should be refactored out I mean.
I think it's missing some stuff from the regexp grammar though (such as lookbehind tokens)